Wrappers/Python/test/test_CIL_CVXPy.py

```diff
@@ -0,0 +1,212 @@
+                        mat[-1,0] = 1
+
+                elif direction == 'backward':
+
+                    # create a sparse matrix with -1 in the -1 and 1 in the main diagonal
+                    mat = 1/discretization[i] * sp.spdiags(np.vstack([-np.ones((1,shape[i])),np.ones((1,shape[i]))]), [-1,0], shape[i], shape[i], format = 'lil')
+
+                    # boundary conditions
+                    if boundaries == 'Neumann':
+                        mat[:,-1] = 0
+                    elif boundaries == 'Periodic':
+                        mat[0,-1] = -1
+
+                # Use Kronecker product to compute the full sparse matrix for the finite difference operator according to the direction.
+                # For a (n x m) array, the forward difference operator in y-direction (x-direction) (with Neumann/Periodic bc) is a (n*m x n*m) sparse array containing -1,1.
+                # Example, for a 3x3 array U, the forward difference operator in y-direction with Neumann bc is a 9x9 sparse array containing -1,1.
+                # To create this sparse matrix, we first create a "kernel" matrix, shown below:
+                # mat = [-1, 1, 0
+                #        0, -1, 1,
+                #        0, 0, 0].
+                # and then use the Kronecker product: allMat[0] = mat x I_m =
+                # matrix([[-1., 1., 0., 0., 0., 0., 0., 0., 0.],
+                #         [ 0., -1., 1., 0., 0., 0., 0., 0., 0.],
+                #         [ 0., 0., 0., 0., 0., 0., 0., 0., 0.],
+                #         [ 0., 0., 0., -1., 1., 0., 0., 0., 0.],
+                #         [ 0., 0., 0., 0., -1., 1., 0., 0., 0.],
+                #         [ 0., 0., 0., 0., 0., 0., 0., 0., 0.],
+                #         [ 0., 0., 0., 0., 0., 0., -1., 1., 0.],
+                #         [ 0., 0., 0., 0., 0., 0., 0., -1., 1.],
+                #         [ 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
+                # where I_m is an (mxm) sparse array with ones in the diagonal.
+                # Then allmat can be applied to a (3x3) array "flatten" columnwise
+                # and represent the forward differences in y direction,i.e.,
+                # [U_{21} - U_{11},
+                #  U_{31} - U_{21},
+                #        0        ,
+                #       ...       ,
+                #       ...       ,
+                #       ...       ,
+                #       ...       ,
+                #       ...       ,
+                #       ...       ]
+
+                # For the x-direction, we have allmat[1] = I_n x mat.
+
+                # For more details, see "Infimal Convolution Regularizations with Discrete l1-type Functionals, S. Setzer, G. Steidl and T. Teuber"
+          
+                # According to the directions, tmpGrad is either a kernel matrix or sparse eye array, which is updated 
+                # using the kronecker product to derive the sparse matrices.
+                if i==0:
```

Comment:
Thanks for adding all the comments, that is really helpful. I now believe I understand. But the part that was the most unclear to me was this one and the next 15 lines or so in which `allmat` is constructed. It wasn't clear that these lines handle any `len_shape`, ie, 2D, 3D, any D, which is very neat. The comments should make that clear. Maybe could you add
```
# The following constructs the finite complete difference matrix for all (len_shape) dimensions and store the ith finite 
# difference matrix in allmat[i].  In 2D we have
# allmat[0] = D kron I
# allmat[1] = I kron D
# and in 3D
# allmat[0] = D kron I kron I
# allmat[1] = I kron D kron I
# allmat[2] = I kron I kron D
# and so on, for kron meaning the kronecker product.
```

Reply:
Improved text and added your comments above.